### PR TITLE
internal: document NameClass and NameRefClass

### PR DESCRIPTION
**Pull request by matklad**
bors r+
🤖

### CODE REVIEW

**Comment by matklad** [comment]
cc @Veykril, this documents what I tried to explain in https://github.com/rust-analyzer/rust-analyzer/issues/7524#issuecomment-877781162

File: crates/ide_db/src/defs.rs [hunk 1]
```diff
@@ -97,13 +97,25 @@ impl Definition {
     }
 }
 
+/// On a first blush, a single `ast::Name` defines a single definition at some
+/// scope. That is, that, by just looking at the syntactical category, we can
+/// unambiguously define the semantic category.
+///
+/// Sadly, that's not 100% true, there are special cases. To make sure that call
```
Comment (lnicola) [inline crates/ide_db/src/defs.rs:104]:
```suggestion
/// Sadly, that's not 100% true, there are special cases. To make sure that callers
```

File: crates/ide_db/src/defs.rs [hunk 2]
```diff
@@ -97,13 +97,25 @@ impl Definition {
...
+/// the code handles all the special cases correctly via exhaustive matching, we
```
Comment (lnicola) [inline crates/ide_db/src/defs.rs:105]:
```suggestion
/// handle all the special cases correctly via exhaustive matching, we
```

**Pull request merged**